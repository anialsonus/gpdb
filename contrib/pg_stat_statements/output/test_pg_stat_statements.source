CREATE TABLE t(a int, b text) DISTRIBUTED BY (a);
-- Known issue: query is not added to pg_stat_statements statistics in
-- case it is planned by GPORCA. So disable GPORCA during tests.
SET optimizer='off';
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

SELECT GROUPING (a) FROM t GROUP BY ROLLUP(a, b);
 grouping 
----------
(0 rows)

-- launch not equivalent query
SELECT GROUPING (b) FROM t GROUP BY ROLLUP(a, b);
 grouping 
----------
(0 rows)

-- check that 2 queries have separate entries
SELECT query, calls FROM pg_stat_statements ORDER BY query;
                       query                       | calls 
---------------------------------------------------+-------
 SELECT GROUPING (a) FROM t GROUP BY ROLLUP(a, b); |     1
 SELECT GROUPING (b) FROM t GROUP BY ROLLUP(a, b); |     1
 SELECT pg_stat_statements_reset();                |     1
(3 rows)

-- check that different grouping options result in separate entries
SELECT COUNT (*) FROM t GROUP BY ROLLUP(a, b);
 count 
-------
(0 rows)

SELECT COUNT (*) FROM t GROUP BY CUBE(a, b);
 count 
-------
(0 rows)

SELECT COUNT (*) FROM t GROUP BY GROUPING SETS(a, b);
 count 
-------
(0 rows)

SELECT COUNT (*) FROM t GROUP BY GROUPING SETS((a), (a, b));
 count 
-------
(0 rows)

SELECT COUNT (*) FROM t GROUP BY a, b;
 count 
-------
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY query;
                            query                             | calls 
--------------------------------------------------------------+-------
 SELECT COUNT (*) FROM t GROUP BY a, b;                       |     1
 SELECT COUNT (*) FROM t GROUP BY CUBE(a, b);                 |     1
 SELECT COUNT (*) FROM t GROUP BY GROUPING SETS((a), (a, b)); |     1
 SELECT COUNT (*) FROM t GROUP BY GROUPING SETS(a, b);        |     1
 SELECT COUNT (*) FROM t GROUP BY ROLLUP(a, b);               |     1
 SELECT GROUPING (a) FROM t GROUP BY ROLLUP(a, b);            |     1
 SELECT GROUPING (b) FROM t GROUP BY ROLLUP(a, b);            |     1
 SELECT pg_stat_statements_reset();                           |     1
 SELECT query, calls FROM pg_stat_statements ORDER BY query;  |     1
(9 rows)

-- check several parameters options in ROLLUP
-- all should result in separate entries
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

SELECT COUNT (*) FROM t GROUP BY ROLLUP(a, b);
 count 
-------
(0 rows)

SELECT COUNT (*) FROM t GROUP BY ROLLUP(b);
 count 
-------
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY query;
                     query                      | calls 
------------------------------------------------+-------
 SELECT COUNT (*) FROM t GROUP BY ROLLUP(a, b); |     1
 SELECT COUNT (*) FROM t GROUP BY ROLLUP(b);    |     1
 SELECT pg_stat_statements_reset();             |     1
(3 rows)

-- check group_id() in query
SELECT group_id() AS c FROM t GROUP BY GROUPING SETS ((a), (a,b));
 c 
---
(0 rows)

-- check anytable parameter for a function
CREATE FUNCTION multiset_1(a anytable) RETURNS TABLE(a int, b text)
    AS '@abs_srcdir@/../../src/test/regress/regress.so', 'multiset_example' LANGUAGE C READS SQL DATA;
CREATE FUNCTION multiset_2(a anytable) RETURNS TABLE(a int, b text)
    AS '@abs_srcdir@/../../src/test/regress/regress.so', 'multiset_example' LANGUAGE C READS SQL DATA;
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

SELECT * FROM multiset_1(TABLE(SELECT * FROM t));
 a | b 
---+---
(0 rows)

SELECT * FROM multiset_1(TABLE(SELECT * FROM t WHERE a=0));
 a | b 
---+---
(0 rows)

SELECT * FROM multiset_2(TABLE(SELECT * FROM t));
 a | b 
---+---
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY query;
                            query                            | calls 
-------------------------------------------------------------+-------
 SELECT * FROM multiset_1(TABLE(SELECT * FROM t));           |     1
 SELECT * FROM multiset_1(TABLE(SELECT * FROM t WHERE a=?)); |     1
 SELECT * FROM multiset_2(TABLE(SELECT * FROM t));           |     1
 SELECT pg_stat_statements_reset();                          |     1
(4 rows)

RESET optimizer;
DROP TABLE t;
