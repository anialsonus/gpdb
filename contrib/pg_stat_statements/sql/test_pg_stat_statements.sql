-- start_ignore
\! gpconfig -c shared_preload_libraries -v 'pg_stat_statements';
\! gpstop -raq -M fast;
\c
CREATE EXTENSION IF NOT EXISTS pg_stat_statements;
DROP TABLE IF EXISTS table_test_pg_stat_statements;
-- end_ignore
-- simple test to check there is no warnings during jumbling of the query
CREATE TABLE table_test_pg_stat_statements
(
    item1 integer,
    item2 integer,
    item3 integer,
    item4 integer
)
DISTRIBUTED BY (item1);

-- Known issue: query is not added to pg_stat_statements statistics in
-- case it is planned by GPORCA. So disable GPORCA during tests.
SET optimizer='off';

SELECT pg_stat_statements_reset();

-- launch 2 equivalent queries  
SELECT
    GROUPING(item1)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 0
GROUP BY ROLLUP(item1, item2);

SELECT
    GROUPING(item1)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 1
GROUP BY ROLLUP(item1, item2);

-- pg_stat_statements statistics should have 2 calls for 1 entry
-- corresponding to the two preceding SELECT queries
SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;

-- launch not equivalent query
SELECT
    GROUPING(item2)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 0
GROUP BY ROLLUP(item1, item2);

-- check that it has separate entry
SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;

-- check that different grouping options result in separate entries
SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item1, item2, item3, item4);

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY CUBE(item1, item2, item3, item4);

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY GROUPING SETS(item1, item2, item3, item4);

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY GROUPING SETS((item1, item2), (item3, item4));

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY item1, item2, item3, item4;

SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;

-- check several parameters options in ROLLUP
-- all should result in separate entries
SELECT pg_stat_statements_reset();

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item1, item2);

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item2, item3);

SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;

RESET optimizer;

DROP TABLE table_test_pg_stat_statements;
-- start_ignore
\! gpconfig -r shared_preload_libraries;
\! gpstop -raq -M fast;
-- end_ignore
