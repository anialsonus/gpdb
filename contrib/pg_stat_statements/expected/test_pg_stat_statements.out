-- simple test to check there is no warnings during jumbling of the query
CREATE TABLE table_test_pg_stat_statements
(
    item1 integer,
    item2 integer,
    item3 integer,
    item4 integer
)
DISTRIBUTED BY (item1);
-- Known issue: query is not added to pg_stat_statements statistics in
-- case it is planned by GPORCA. So disable GPORCA during tests.
SET optimizer='off';
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

-- launch 2 equivalent queries  
SELECT
    GROUPING(item1)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 0
GROUP BY ROLLUP(item1, item2);
 grouping 
----------
(0 rows)

SELECT
    GROUPING(item1)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 1
GROUP BY ROLLUP(item1, item2);
 grouping 
----------
(0 rows)

-- pg_stat_statements statistics should have 2 calls for 1 entry
-- corresponding to the two preceding SELECT queries
SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;
               query                | calls 
------------------------------------+-------
 SELECT                            +|     2
     GROUPING(item1)               +| 
 FROM                              +| 
     table_test_pg_stat_statements +| 
 WHERE                             +| 
     item3 = ?                     +| 
 GROUP BY ROLLUP(item1, item2);     | 
 SELECT pg_stat_statements_reset(); |     1
(2 rows)

-- launch not equivalent query
SELECT
    GROUPING(item2)
FROM
    table_test_pg_stat_statements
WHERE
    item3 = 0
GROUP BY ROLLUP(item1, item2);
 grouping 
----------
(0 rows)

-- check that it has separate entry
SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;
                            query                            | calls 
-------------------------------------------------------------+-------
 SELECT                                                     +|     2
     GROUPING(item1)                                        +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 WHERE                                                      +| 
     item3 = ?                                              +| 
 GROUP BY ROLLUP(item1, item2);                              | 
 SELECT                                                     +|     1
     GROUPING(item2)                                        +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 WHERE                                                      +| 
     item3 = ?                                              +| 
 GROUP BY ROLLUP(item1, item2);                              | 
 SELECT pg_stat_statements_reset();                          |     1
 SELECT query, calls FROM pg_stat_statements ORDER BY QUERY; |     1
(4 rows)

-- check that different grouping options result in separate entries
SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item1, item2, item3, item4);
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY CUBE(item1, item2, item3, item4);
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY GROUPING SETS(item1, item2, item3, item4);
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY GROUPING SETS((item1, item2), (item3, item4));
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY item1, item2, item3, item4;
 count 
-------
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;
                            query                            | calls 
-------------------------------------------------------------+-------
 SELECT                                                     +|     1
     COUNT (*)                                              +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 GROUP BY CUBE(item1, item2, item3, item4);                  | 
 SELECT                                                     +|     1
     COUNT (*)                                              +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 GROUP BY GROUPING SETS((item1, item2), (item3, item4));     | 
 SELECT                                                     +|     1
     COUNT (*)                                              +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 GROUP BY GROUPING SETS(item1, item2, item3, item4);         | 
 SELECT                                                     +|     1
     COUNT (*)                                              +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 GROUP BY item1, item2, item3, item4;                        | 
 SELECT                                                     +|     1
     COUNT (*)                                              +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 GROUP BY ROLLUP(item1, item2, item3, item4);                | 
 SELECT                                                     +|     2
     GROUPING(item1)                                        +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 WHERE                                                      +| 
     item3 = ?                                              +| 
 GROUP BY ROLLUP(item1, item2);                              | 
 SELECT                                                     +|     1
     GROUPING(item2)                                        +| 
 FROM                                                       +| 
     table_test_pg_stat_statements                          +| 
 WHERE                                                      +| 
     item3 = ?                                              +| 
 GROUP BY ROLLUP(item1, item2);                              | 
 SELECT pg_stat_statements_reset();                          |     1
 SELECT query, calls FROM pg_stat_statements ORDER BY QUERY; |     2
(9 rows)

-- check several parameters options in ROLLUP
-- all should result in separate entries
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item1, item2);
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY ROLLUP(item2, item3);
 count 
-------
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;
               query                | calls 
------------------------------------+-------
 SELECT                            +|     1
     COUNT (*)                     +| 
 FROM                              +| 
     table_test_pg_stat_statements +| 
 GROUP BY ROLLUP(item1, item2);     | 
 SELECT                            +|     1
     COUNT (*)                     +| 
 FROM                              +| 
     table_test_pg_stat_statements +| 
 GROUP BY ROLLUP(item2, item3);     | 
 SELECT pg_stat_statements_reset(); |     1
(3 rows)

-- check several options in simple GROUP BY
-- all should result in separate entries
SELECT pg_stat_statements_reset();
 pg_stat_statements_reset 
--------------------------
 
(1 row)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY item1, item2;
 count 
-------
(0 rows)

SELECT
    COUNT (*)
FROM
    table_test_pg_stat_statements
GROUP BY item2, item3;
 count 
-------
(0 rows)

SELECT query, calls FROM pg_stat_statements ORDER BY QUERY;
               query                | calls 
------------------------------------+-------
 SELECT                            +|     1
     COUNT (*)                     +| 
 FROM                              +| 
     table_test_pg_stat_statements +| 
 GROUP BY item1, item2;             | 
 SELECT                            +|     1
     COUNT (*)                     +| 
 FROM                              +| 
     table_test_pg_stat_statements +| 
 GROUP BY item2, item3;             | 
 SELECT pg_stat_statements_reset(); |     1
(3 rows)

RESET optimizer;
DROP TABLE table_test_pg_stat_statements;
